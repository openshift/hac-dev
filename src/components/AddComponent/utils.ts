import * as React from 'react';
import { useFormikContext } from 'formik';
import { useK8sWatchResource } from '../../dynamic-plugin-sdk';
import {
  ComponentDetectionQueryGroupVersionKind,
  SPIAccessTokenBindingGroupVersionKind,
} from '../../models';
import {
  ComponentDetectionQueryKind,
  DetectedComponents,
  SPIAccessTokenBindingKind,
  SPIAccessTokenBindingPhase,
} from '../../types';
import { createComponentDetectionQuery } from '../../utils/create-utils';
import { NamespaceContext } from './../NamespacedPage/NamespacedPage';

/**
 * Watch the SPIAccessTokenBinding resource and open auth window when provided.
 * Upon successful injection, set the specified secret.
 */
export const useAccessTokenBindingAuth = (name: string) => {
  const { namespace } = React.useContext(NamespaceContext);
  const { setFieldValue } = useFormikContext();
  const [binding, loaded] = useK8sWatchResource<SPIAccessTokenBindingKind>({
    groupVersionKind: SPIAccessTokenBindingGroupVersionKind,
    name,
    namespace,
  });

  React.useEffect(() => {
    if (!name || !loaded) return;
    const oAuthUrl = binding.status?.oAuthUrl;
    if (oAuthUrl) {
      window.open(oAuthUrl);
    }
  }, [binding?.status?.oAuthUrl, loaded, name]);

  React.useEffect(() => {
    if (!name || !loaded) return;
    if (binding.status?.phase === SPIAccessTokenBindingPhase.Injected) {
      setFieldValue('git.authSecret', binding.status.syncedObjectRef.name);
      // eslint-disable-next-line no-console
      console.log('Git repository successfully authorized.');
    } else if (binding.status?.phase === SPIAccessTokenBindingPhase.Error) {
      // eslint-disable-next-line no-console
      console.log('Error in binding status ', binding.status.errorMessage);
    }
  }, [
    name,
    loaded,
    setFieldValue,
    binding?.status?.phase,
    binding?.status?.errorMessage,
    binding?.status?.syncedObjectRef?.name,
  ]);
};

/**
 * Create a ComponentDetectionQuery when any of the params change,
 * and return the detected components when detection is completed.
 */
export const useComponentDetection = (
  source: string,
  application: string,
  isMultiComponent?: boolean,
  authSecret?: string,
): [DetectedComponents, any] => {
  const { namespace } = React.useContext(NamespaceContext);
  const [cdqName, setCdqName] = React.useState<string>();
  const [createError, setCreateError] = React.useState();

  const [cdq, loaded, loadError] = useK8sWatchResource<ComponentDetectionQueryKind>(
    cdqName
      ? {
          groupVersionKind: ComponentDetectionQueryGroupVersionKind,
          name: cdqName,
          namespace,
          isList: false,
        }
      : {},
  );

  React.useEffect(() => {
    setCdqName(null);
    setCreateError(null);
    if (source) {
      createComponentDetectionQuery(application, source, namespace, isMultiComponent, authSecret)
        .then((resource) => setCdqName(resource.metadata.name))
        .catch(setCreateError);
    }
  }, [application, authSecret, isMultiComponent, namespace, source]);

  const detectedComponents = React.useMemo(() => {
    if (cdqName && loaded && cdq) {
      if (cdqName === cdq.metadata.name) {
        return cdq?.status?.componentDetected;
      }
    }
  }, [cdqName, cdq, loaded]);

  const error = React.useMemo(() => {
    if (createError) {
      return createError;
    }
    if (cdqName) {
      if (loadError) {
        return loadError;
      }

      const completeCondition = cdq?.status?.conditions?.find(
        (condition) => condition.type === 'Completed',
      );

      if (loaded && !detectedComponents && completeCondition) {
        if (cdqName === cdq.metadata.name) {
          return completeCondition.message;
        }
      }
    }
  }, [cdq, cdqName, createError, detectedComponents, loadError, loaded]);

  return [detectedComponents, error];
};
