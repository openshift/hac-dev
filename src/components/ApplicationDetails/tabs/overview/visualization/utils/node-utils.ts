import { K8sResourceCommon } from '@openshift/dynamic-plugin-sdk-utils';
import { Node, PipelineNodeModel, RunStatus } from '@patternfly/react-topology';
import { PipelineRunLabel } from '../../../../../../consts/pipelinerun';
import { pipelineRunFilterReducer, pipelineRunStatus, runStatus } from '../../../../../../shared';
import { PipelineRunKind } from '../../../../../../shared/components/pipeline-run-logs/types';
import { ComponentKind } from '../../../../../../types';
import { GitOpsDeploymentHealthStatus } from '../../../../../../types/gitops-deployment';
import { hasPACProvisionDone } from '../../../../../Components/BuildStatusColumn';
import { DEFAULT_NODE_HEIGHT } from '../../../../../topology/const';
import { NodeType } from '../const';
import { WorkflowNodeModel, WorkflowNodeModelData, WorkflowNodeType } from '../types';
import { getNodeWidth } from './visualization-utils';

const UNKNOWN_STATUS = 'unknown';
export const NEEDS_MERGE_STATUS = 'PR needs merge';

const RUN_STATUS_SEVERITIES = [
  UNKNOWN_STATUS,
  runStatus.Succeeded,
  runStatus.Idle,
  runStatus.Skipped,
  runStatus.Pending,
  runStatus['In Progress'],
  runStatus.PipelineNotStarted,
  runStatus.Running,
  NEEDS_MERGE_STATUS,
  runStatus.Cancelled,
  runStatus.FailedToStart,
  runStatus.Failed,
];

const COMPONENT_DESC =
  'A component is an image built from code in a source repository. Applications are sets of components that run together on environments.';
const BUILD_DESC = `Every component requires a build in order to deploy it to an environment. Whenever you create a new component, we send a pull request to merge our build pipeline to your component's source code`;
const TESTS_DESC = `An integration test is a test pipeline you build and add to the application CI/CD definition. It tests each component as itâ€™s added and then tests the entire application as a whole.`;
const STATIC_ENV_DESC = `A static environment is a set of compute resources bundled together. Use static environments for developing, testing, and staging before releasing your application. You can share static environments across all applications within the workspace.`;
const RELEASE_DESC = `After pushing your application to release, your application goes through a series of tests through the release pipeline to ensure the application complies with the  release policy set on the release target, also known as the "managed environment".`;
const MANAGED_ENV_DESC = `A managed environment is your application release target. This release target is an external environment, set up in an external workspace, and managed by another team. It is set for each application and not automatically shared among applications within the workspace.`;

export const TYPE_DESCRIPTIONS = {
  [WorkflowNodeType.COMPONENT]: COMPONENT_DESC,
  [WorkflowNodeType.BUILD]: BUILD_DESC,
  [WorkflowNodeType.TESTS]: TESTS_DESC,
  [WorkflowNodeType.COMPONENT_TEST]: TESTS_DESC,
  [WorkflowNodeType.APPLICATION_TEST]: TESTS_DESC,
  [WorkflowNodeType.STATIC_ENVIRONMENT]: STATIC_ENV_DESC,
  [WorkflowNodeType.RELEASE]: RELEASE_DESC,
  [WorkflowNodeType.MANAGED_ENVIRONMENT]: MANAGED_ENV_DESC,
};

export const statusToRunStatus = (status: string): RunStatus => {
  switch (status) {
    case runStatus.Succeeded:
    case GitOpsDeploymentHealthStatus.Healthy:
      return RunStatus.Succeeded;
    case runStatus.Failed:
    case runStatus.FailedToStart:
    case GitOpsDeploymentHealthStatus.Degraded:
      return RunStatus.Failed;
    case runStatus.Running:
    case GitOpsDeploymentHealthStatus.Progressing:
    case runStatus['In Progress']:
      return RunStatus.Running;
    case NEEDS_MERGE_STATUS:
      return RunStatus.Cancelled; // to show a warning
    case RunStatus.Pending:
    case GitOpsDeploymentHealthStatus.Suspended:
    case GitOpsDeploymentHealthStatus.Missing:
    case GitOpsDeploymentHealthStatus.Unknown:
      return RunStatus.Pending;
    default:
      return undefined;
  }
};

export const getLinkDataForElement = (
  element: Node<PipelineNodeModel, WorkflowNodeModelData>,
): { tab?: string; path?: string; filter?: { name: string; value: string } } => {
  const { workflowType, label, isDisabled, groupNode, status } = element.getData();

  switch (workflowType) {
    case WorkflowNodeType.COMPONENT:
      return {
        tab: 'components',
        filter: !groupNode && !isDisabled ? { name: 'name', value: label } : undefined,
      };
    case WorkflowNodeType.BUILD:
      if (status === NEEDS_MERGE_STATUS) {
        return {
          tab: 'components',
          filter:
            !groupNode && !isDisabled && label
              ? { name: 'name', value: label.replace('Build for ', '') }
              : undefined,
        };
      }
      return {
        tab: 'pipelineruns',
        // TODO: filter by build once the PLR tab supports filtering
      };
    case WorkflowNodeType.TESTS:
    case WorkflowNodeType.COMPONENT_TEST:
    case WorkflowNodeType.APPLICATION_TEST:
      return !groupNode && !isDisabled
        ? {
            path: `/stonesoup/${element.getData().application}/integrationtests/${label}`,
          }
        : {
            tab: 'integrationtests',
          };
    case WorkflowNodeType.STATIC_ENVIRONMENT:
      return {
        tab: 'environments',
        filter:
          !groupNode && !isDisabled
            ? { name: 'name', value: label }
            : { name: 'envType', value: 'static' },
      };
    case WorkflowNodeType.MANAGED_ENVIRONMENT:
    case WorkflowNodeType.RELEASE:
      return {
        tab: 'environments',
        filter:
          !groupNode && !isDisabled
            ? { name: 'name', value: label }
            : { name: 'envType', value: 'managed' },
      };
    default:
      return {
        tab: 'overview',
      };
  }
};

export const getLinksForElement = (
  element: Node<PipelineNodeModel, WorkflowNodeModelData>,
): { elementRef: string; pipelinesRef: string } => {
  const linkData = getLinkDataForElement(element);

  const appPath = `/stonesoup/applications/${element.getData().application}`;
  const tabPath = linkData.tab ? `/${linkData.tab}` : '';
  const filter = linkData.filter ? `?${linkData.filter.name}=${linkData.filter.value}` : '';

  return {
    elementRef: linkData.path ? linkData.path : `${appPath}${tabPath}${filter}`,
    pipelinesRef: `${appPath}/pipelineruns`,
  };
};

export const getRunStatusComponent = (
  component: ComponentKind,
  pipelineRuns: PipelineRunKind[],
) => {
  const latestPipelineRun = pipelineRuns
    .filter((pr) => pr.metadata.labels?.[PipelineRunLabel.COMPONENT] === component.metadata.name)
    .sort(
      (a, b) =>
        new Date(b.metadata.creationTimestamp).getTime() -
        new Date(a.metadata.creationTimestamp).getTime(),
    )?.[0];
  if (latestPipelineRun) {
    return pipelineRunFilterReducer(latestPipelineRun);
  } else if (hasPACProvisionDone(component)) {
    return NEEDS_MERGE_STATUS;
  }
  return UNKNOWN_STATUS;
};

export const worstWorkflowStatus = (workFlows: PipelineNodeModel[]) =>
  workFlows.reduce((worstStatus, c) => {
    const statusSeverity = RUN_STATUS_SEVERITIES.indexOf(c.data.status) || 0;
    if (statusSeverity > RUN_STATUS_SEVERITIES.indexOf(worstStatus)) {
      return c.data.status;
    }
    return worstStatus;
  }, '');

export const resourceToPipelineNode = (
  resource: K8sResourceCommon,
  application: string,
  workflowType: WorkflowNodeType,
  runAfterTasks: string[] = [],
  status?: runStatus | string,
  label?: string,
): WorkflowNodeModel<WorkflowNodeModelData> => ({
  id: resource.metadata.uid,
  label: label || resource.metadata.name,
  type: NodeType.WORKFLOW_NODE,
  height: DEFAULT_NODE_HEIGHT,
  width: getNodeWidth(resource.metadata.name, status),
  runAfterTasks,
  data: {
    application,
    label: label || resource.metadata.name,
    workflowType,
    status,
    resources: [resource],
  },
});

export const emptyPipelineNode = (
  id: string,
  application: string,
  label: string,
  workflowType: WorkflowNodeType,
  runAfterTasks: string[] = [],
): WorkflowNodeModel<WorkflowNodeModelData> => ({
  id,
  label,
  type: NodeType.WORKFLOW_NODE,
  height: DEFAULT_NODE_HEIGHT,
  width: getNodeWidth(label),
  runAfterTasks,
  data: {
    application,
    label,
    isDisabled: true,
    workflowType,
  },
});

export const groupToPipelineNode = (
  id: string,
  application: string,
  label: string,
  workflowType: WorkflowNodeType,
  runAfterTasks: string[] = [],
  group: boolean,
  children?: string[],
  childNodes?: WorkflowNodeModel<WorkflowNodeModelData>[],
  resources?: K8sResourceCommon[],
  status?: runStatus | string,
): WorkflowNodeModel<WorkflowNodeModelData> => {
  const isDisabled = !resources?.length;
  return {
    id,
    label,
    height: DEFAULT_NODE_HEIGHT,
    type: group ? NodeType.WORKFLOW_GROUP : NodeType.WORKFLOW_NODE,
    width: getNodeWidth(label, status, group ? undefined : children),
    group,
    children,
    runAfterTasks: group ? [] : runAfterTasks,
    style: { padding: [35] },
    data: {
      application,
      label,
      workflowType,
      isDisabled,
      groupNode: !group,
      status,
      resources,
      children: !isDisabled ? childNodes : undefined,
    },
  };
};

export const getBuildNodeForComponent = (
  component: ComponentKind,
  application: string,
  latestBuilds: PipelineRunKind[],
): WorkflowNodeModel<WorkflowNodeModelData> => {
  const latestBuild = latestBuilds.find(
    (build) => component.metadata.name === build.metadata.labels?.[PipelineRunLabel.COMPONENT],
  );
  if (latestBuild) {
    return resourceToPipelineNode(
      latestBuild,
      application,
      WorkflowNodeType.BUILD,
      [component.metadata.uid],
      pipelineRunStatus(latestBuild),
      `Build for ${component.metadata.name}`,
    );
  }
  return {
    id: `${component.metadata.uid}-missing`,
    label: `Build for ${component.metadata.name}`,
    type: NodeType.WORKFLOW_NODE,
    height: DEFAULT_NODE_HEIGHT,
    width: 0,
    runAfterTasks: [component.metadata.uid],
    data: {
      application,
      label: `Build for ${component.metadata.name}`,
      isDisabled: false,
      status: NEEDS_MERGE_STATUS,
      workflowType: WorkflowNodeType.BUILD,
    },
  };
};
