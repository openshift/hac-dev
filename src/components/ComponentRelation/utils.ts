import { k8sGetResource, k8sUpdateResource } from '@openshift/dynamic-plugin-sdk-utils';
import { union } from 'lodash-es';
import * as yup from 'yup';
import { ComponentModel } from '../../models';
import { ComponentKind, NudgeStats } from '../../types';
import {
  ComponentRelationFormikValue,
  ComponentRelationNudgeType,
  ComponentRelationValue,
} from './type';

export const DUPLICATE_RELATONSHIP = 'duplicate-component-relationship';

export const componentRelationValidationSchema = yup.mixed().test(
  (values: ComponentRelationFormikValue) =>
    yup
      .object()
      .shape({
        relations: yup
          .array()
          .of(
            yup
              .object()
              .shape({
                source: yup.string().required(),

                nudgeType: yup.string().required(),
                target: yup.array().of(yup.string()).required().min(0),
              })
              .test('duplicate-relation-test', DUPLICATE_RELATONSHIP, (value) => {
                const filteredValue = values.relations.filter(
                  (val) => val.source === value.source && val.nudgeType === value.nudgeType,
                );
                return filteredValue.length < 2;
              }),
          )
          .required()
          .min(1),
      })
      .validate(values, { abortEarly: false }) as any,
);

export const transformNudgeData = (data: ComponentRelationValue[]): { [key: string]: string[] } => {
  return data.reduce((acc, { source, nudgeType, target }) => {
    if (!source) return acc;
    if (nudgeType === ComponentRelationNudgeType.NUDGES) {
      acc[source] = acc[source] ? union(acc[source], target) : target;
      return acc;
    }
    target.length > 0 &&
      target.forEach((t) => {
        acc[t] = acc[t] ? union(acc[t], [source]) : [source];
      });
    return acc;
  }, {});
};

const mergeNudgeDataIntoResource = (resource: ComponentKind, nudgeData: string[]) => {
  return {
    ...resource,
    spec: {
      ...resource.spec,
      [NudgeStats.NUDGES]: nudgeData,
    },
  };
};

export const updateNudgeDependencies = async (
  values: ComponentRelationValue[],
  namespace: string,
  dryRun?: boolean,
) => {
  const transformedData = transformNudgeData(values);
  const data = [];
  for (const [componentName, nudgeData] of Object.entries(transformedData)) {
    // fetch new copy because component dependency might have been updated for this component
    const component = await k8sGetResource<ComponentKind>({
      model: ComponentModel,
      queryOptions: { name: componentName, ns: namespace },
    });
    const resource = mergeNudgeDataIntoResource(component, nudgeData);
    const result = await k8sUpdateResource({
      model: ComponentModel,
      resource,
      queryOptions: {
        name: component.metadata.name,
        ns: component.metadata.namespace,
        ...(dryRun && { queryParams: { dryRun: 'All' } }),
      },
    });
    data.push(result);
  }
  return data;
};
