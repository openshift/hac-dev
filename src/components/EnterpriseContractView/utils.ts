import { K8sResourceCommon } from '@openshift/dynamic-plugin-sdk-utils';

const ENTERPRISE_CONTRACT_LABEL = 'build.appstudio.redhat.com/pipeline';

export const isResourceEnterpriseContract = (resource: K8sResourceCommon): boolean => {
  return resource?.metadata?.labels[ENTERPRISE_CONTRACT_LABEL] === 'enterprise-contract';
};
