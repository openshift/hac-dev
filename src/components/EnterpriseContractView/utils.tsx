import * as React from 'react';
import { K8sResourceCommon } from '@openshift/dynamic-plugin-sdk-utils';
import CheckCircleIcon from '@patternfly/react-icons/dist/js/icons/check-circle-icon';
import DotCircleIcon from '@patternfly/react-icons/dist/js/icons/dot-circle-icon';
import ExclamationCircleIcon from '@patternfly/react-icons/dist/js/icons/exclamation-circle-icon';
import ExclamationTriangleIcon from '@patternfly/react-icons/dist/js/icons/exclamation-triangle-icon';
import { global_danger_color_100 as redColor } from '@patternfly/react-tokens/dist/js/global_danger_color_100';
import { global_success_color_100 as greenColor } from '@patternfly/react-tokens/dist/js/global_success_color_100';
import { global_warning_color_100 as yellowColor } from '@patternfly/react-tokens/dist/js/global_warning_color_100';
import { ENTERPRISE_CONTRACT_STATUS, EnterpriseContractResult } from './types';

const ENTERPRISE_CONTRACT_LABEL = 'build.appstudio.redhat.com/pipeline';

export const isResourceEnterpriseContract = (resource: K8sResourceCommon): boolean => {
  return resource?.metadata?.labels[ENTERPRISE_CONTRACT_LABEL] === 'enterprise-contract';
};

export const getRuleStatus = (type: ENTERPRISE_CONTRACT_STATUS) => {
  switch (type) {
    case ENTERPRISE_CONTRACT_STATUS.successes:
      return (
        <>
          <CheckCircleIcon color={greenColor.value} /> {ENTERPRISE_CONTRACT_STATUS.successes}
        </>
      );
    case ENTERPRISE_CONTRACT_STATUS.violations:
      return (
        <>
          <ExclamationCircleIcon color={redColor.value} /> {ENTERPRISE_CONTRACT_STATUS.violations}
        </>
      );
    case ENTERPRISE_CONTRACT_STATUS.warnings:
      return (
        <>
          <ExclamationTriangleIcon color={yellowColor.value} />{' '}
          {ENTERPRISE_CONTRACT_STATUS.warnings}
        </>
      );
    default:
      return (
        <>
          <DotCircleIcon /> Missing
        </>
      );
  }
};

/**
 * This regex expect the logs from tekton results to be in this formay
 *
 * ```
 * step-vulnerabilities :-
 * Lorem Ipsum some logs
 *
 * step-report-json :-
 * {"success":true,"components":[{"name":"devfile-sample-code-with-quarkus-1",<... ec report in JSON ...>,}]}
 *
 * ```
 *
 */
const EC_REPORT_JSON_REGEX = /((?<=step-report-json\s*:-\s*)(\{.*?\})(?=\s*step-|$))/g;

export const extractEcResultsFromTaskRunLogs = (logs: string): EnterpriseContractResult => {
  const extractedLogs = logs.match(EC_REPORT_JSON_REGEX);
  try {
    return JSON.parse(extractedLogs[0]);
  } catch {
    return null;
  }
};
