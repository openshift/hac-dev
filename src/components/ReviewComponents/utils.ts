import { ResourceRequirements } from '../../types';
import { ComponentValues } from './types';

const getResourceData = (res: string) => {
  const resourcesRegEx = /^[0-9]*|[a-zA-Z]*/g;
  return res.match(resourcesRegEx);
};

const CPUResourceMap = {
  m: 'millicores',
  '': 'cores',
};

type ResourceData = {
  limits?: { cpu?: string; memory?: string };
  requests?: { cpu?: string; memory?: string };
};

export const createResourceData = (resources: ResourceData) => {
  const memory = (resources?.limits?.memory || resources?.requests?.memory) ?? '512Mi';
  const cpu = (resources?.limits?.cpu || resources?.requests?.cpu) ?? '1';
  const [memoryResource, memoryUnit] = getResourceData(memory);
  const [cpuResource, cpuUnit] = getResourceData(cpu);

  return {
    cpu: cpuResource || '',
    cpuUnit: CPUResourceMap[cpuUnit] || CPUResourceMap[''],
    memory: memoryResource || '',
    memoryUnit: memoryUnit || 'Gi',
  };
};

type FormResources = ComponentValues['resources'];

export const transformResources = (formResources: FormResources): ResourceRequirements => {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars, no-unused-vars
  const cpuUnit = Object.entries(CPUResourceMap).find(([_, v]) => v === formResources.cpuUnit)[0];
  return {
    requests: {
      cpu: `${formResources.cpu}${cpuUnit}`,
      memory: `${formResources.memory}${formResources.memoryUnit}`,
    },
  };
};

export const transformComponentValues = (components) => {
  return components.reduce(
    (acc, val) => ({
      ...acc,
      [val.name]: {
        name: val.name
          .split(/ |\.|:/)
          .join('-')
          .toLowerCase(),
        source: val.data?.source,
        resources: createResourceData(val.data?.resources || {}),
        replicas: val.data?.replicas || 1,
        targetPort: val.data.targetPort || 8080,
        route: val.data.route,
        env: val.data.env,
      },
    }),
    {},
  );
};
