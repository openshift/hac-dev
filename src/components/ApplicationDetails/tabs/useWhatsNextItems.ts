import { useApplicationPipelineGitHubApp } from '../../../hooks/useApplicationPipelineGitHubApp';
import componentsIcon from '../../../imgs/Components.svg';
import editCodeIcon from '../../../imgs/Edit code.svg';
import gitAppIcon from '../../../imgs/git-app.svg';
import integrationTestIcon from '../../../imgs/Integration-test.svg';
import pipelineIcon from '../../../imgs/Pipeline.svg';
import releaseIcon from '../../../imgs/Release.svg';
import { ComponentModel, IntegrationTestScenarioModel, ReleasePlanModel } from '../../../models';
import { useAccessReviewForModel } from '../../../utils/rbac';
import { useWorkspaceInfo } from '../../../utils/workspace-context-utils';
import { createCustomizeAllPipelinesModalLauncher } from '../../CustomizedPipeline/CustomizePipelinesModal';
import { useModalLauncher } from '../../modal/ModalProvider';
import { WhatsNextItem } from '../../WhatsNext/WhatsNextSection';

export const useWhatsNextItems = (applicationName: string) => {
  const showModal = useModalLauncher();
  const { workspace, namespace } = useWorkspaceInfo();
  const { url: githubAppURL } = useApplicationPipelineGitHubApp();
  const [canCreateComponent] = useAccessReviewForModel(ComponentModel, 'create');
  const [canCreateIntegrationTest] = useAccessReviewForModel(
    IntegrationTestScenarioModel,
    'create',
  );
  const [canCreateReleasePlan] = useAccessReviewForModel(ReleasePlanModel, 'create');

  const whatsNextItems: WhatsNextItem[] = [
    {
      title: 'Grow your application',
      description: 'Grow your application by adding components.',
      icon: componentsIcon,
      cta: {
        label: 'Add component',
        href: `/application-pipeline/workspaces/${workspace}/import?application=${applicationName}`,
        disabled: !canCreateComponent,
        disabledTooltip: "You don't have access to add a component",
        testId: 'add-component',
        analytics: {
          link_name: 'add-component',
          link_location: 'whats-next',
          app_name: applicationName,
          workspace,
        },
      },
      helpLink: 'https://konflux-ci.dev/docs/how-tos/creating/',
    },
    {
      title: 'Add integration tests',
      description:
        'Integration tests run in parallel, validating each new component build with the latest version of all other components.',
      icon: integrationTestIcon,
      cta: {
        label: 'Add a test',
        href: `/application-pipeline/workspaces/${workspace}/applications/${applicationName}/integrationtests/add`,
        disabled: !canCreateIntegrationTest,
        disabledTooltip: "You don't have access to add an integration test",
        testId: 'add-test',
        analytics: {
          link_name: 'add-test',
          link_location: 'whats-next',
          app_name: applicationName,
          workspace,
        },
      },
      helpLink: 'https://konflux-ci.dev/docs/how-tos/testing/integration/adding/',
    },
    {
      title: 'Create a release plan',
      description:
        'A release object represents deployable snapshot of your application. To release your code, create a release plan with release details.',
      icon: releaseIcon,
      cta: {
        label: 'Create a plan',
        href: `/application-pipeline/release/workspaces/${workspace}/release-plan/create`,
        disabled: !canCreateReleasePlan,
        disabledTooltip: "You don't have access to create a release plan",
        testId: 'add-release-plan',
        analytics: {
          link_name: 'add-release-plan',
          link_location: 'whats-next',
          app_name: applicationName,
          workspace,
        },
      },
      helpLink: 'https://konflux-ci.dev/docs/advanced-how-tos/releasing/',
    },
    {
      title: 'Install our GitHub app',
      description: 'Install the GitHub app to monitor your work from a commit to deployment.',
      icon: gitAppIcon,
      cta: {
        label: 'Start the flow',
        href: githubAppURL,
        external: true,
        analytics: {
          link_name: 'install-github-app',
          link_location: 'whats-next',
          app_name: applicationName,
          workspace,
        },
      },
      helpLink: 'https://konflux-ci.dev/docs/how-tos/creating/',
    },
    {
      title: 'Make a code change',
      description: 'Make a change to your source code to automatically trigger a new build.',
      icon: editCodeIcon,
      cta: {
        label: 'View build activity',
        href: `/application-pipeline/workspaces/${workspace}/applications/${applicationName}/activity`,
        analytics: {
          link_name: 'view-build-activity',
          link_location: 'whats-next',
          app_name: applicationName,
          workspace,
        },
      },
      helpLink: 'https://konflux-ci.dev/docs/how-tos/creating/',
    },
    {
      title: 'Manage build pipelines',
      description:
        'Add some automation by upgrading your default build pipelines to custom build pipelines.',
      icon: pipelineIcon,
      cta: {
        label: 'Manage build pipelines',
        onClick: () =>
          showModal(createCustomizeAllPipelinesModalLauncher(applicationName, namespace)),
        analytics: {
          link_name: 'manage-build-pipelines',
          link_location: 'whats-next',
          app_name: applicationName,
          workspace,
        },
      },
      helpLink: 'https://konflux-ci.dev/docs/how-tos/configuring/customizing-the-build/',
    },
  ];
  return whatsNextItems;
};
