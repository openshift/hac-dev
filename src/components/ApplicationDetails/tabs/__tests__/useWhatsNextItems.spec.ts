import { renderHook } from '@testing-library/react';
import { useAccessReviewForModel } from '../../../../utils/rbac';
import { useModalLauncher } from '../../../modal/ModalProvider';
import { useWhatsNextItems } from '../useWhatsNextItems';

jest.mock('../../../../utils/workspace-context-utils', () => ({
  useWorkspaceInfo: jest.fn(() => ({ namespace: 'test-ns' })),
}));

jest.mock('../../../../hooks/useApplicationPipelineGitHubApp', () => ({
  useApplicationPipelineGitHubApp: jest.fn(() => ({
    name: 'test-app',
    url: 'https://github.com/test-app',
  })),
}));

jest.mock('../../../../utils/rbac', () => ({
  useAccessReviewForModel: jest.fn(() => [true, true]),
}));

jest.mock('../../../modal/ModalProvider', () => ({
  useModalLauncher: jest.fn(() => () => {}),
}));

describe('useWhatsNextItems', () => {
  it('should return a list of whats next items', () => {
    const applicationName = 'test-application';
    const { result } = renderHook(() => useWhatsNextItems(applicationName));
    expect(result.current).toHaveLength(6);
    expect(result.current).toMatchSnapshot();
  });

  it('should return a list of whats next items with disabled add component button', () => {
    const applicationName = 'test-application';
    (useAccessReviewForModel as jest.Mock).mockReturnValueOnce([false, true]);
    const { result } = renderHook(() => useWhatsNextItems(applicationName));
    expect(result.current[0].cta.disabled).toBe(true);
  });

  it('should launch modal for manage build pipelines', () => {
    const applicationName = 'test-application';
    const showModal = jest.fn();
    (useModalLauncher as jest.Mock).mockReturnValueOnce(showModal);
    const { result } = renderHook(() => useWhatsNextItems(applicationName));
    result.current[5].cta.onClick();
    expect(showModal).toHaveBeenCalled();
  });
});
