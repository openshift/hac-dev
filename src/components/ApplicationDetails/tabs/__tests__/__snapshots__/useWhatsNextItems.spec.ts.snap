// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useWhatsNextItems should return a list of whats next items 1`] = `
Array [
  Object {
    "cta": Object {
      "analytics": Object {
        "app_name": "test-application",
        "link_location": "whats-next",
        "link_name": "add-component",
        "workspace": undefined,
      },
      "disabled": false,
      "disabledTooltip": "You don't have access to add a component",
      "href": "/application-pipeline/workspaces/undefined/import?application=test-application",
      "label": "Add component",
      "testId": "add-component",
    },
    "description": "Grow your application by adding components.",
    "helpLink": "https://konflux-ci.dev/docs/how-tos/creating/",
    "icon": "test-file-stub",
    "title": "Grow your application",
  },
  Object {
    "cta": Object {
      "analytics": Object {
        "app_name": "test-application",
        "link_location": "whats-next",
        "link_name": "add-test",
        "workspace": undefined,
      },
      "disabled": false,
      "disabledTooltip": "You don't have access to add an integration test",
      "href": "/application-pipeline/workspaces/undefined/applications/test-application/integrationtests/add",
      "label": "Add a test",
      "testId": "add-test",
    },
    "description": "Integration tests run in parallel, validating each new component build with the latest version of all other components.",
    "helpLink": "https://konflux-ci.dev/docs/how-tos/testing/integration/adding/",
    "icon": "test-file-stub",
    "title": "Add integration tests",
  },
  Object {
    "cta": Object {
      "analytics": Object {
        "app_name": "test-application",
        "link_location": "whats-next",
        "link_name": "add-release-plan",
        "workspace": undefined,
      },
      "disabled": false,
      "disabledTooltip": "You don't have access to create a release plan",
      "href": "/application-pipeline/release/workspaces/undefined/release-plan/create",
      "label": "Create a plan",
      "testId": "add-release-plan",
    },
    "description": "A release object represents deployable snapshot of your application. To release your code, create a release plan with release details.",
    "helpLink": "https://konflux-ci.dev/docs/advanced-how-tos/releasing/",
    "icon": "test-file-stub",
    "title": "Create a release plan",
  },
  Object {
    "cta": Object {
      "analytics": Object {
        "app_name": "test-application",
        "link_location": "whats-next",
        "link_name": "install-github-app",
        "workspace": undefined,
      },
      "external": true,
      "href": "https://github.com/test-app",
      "label": "Start the flow",
    },
    "description": "Install the GitHub app to monitor your work from a commit to deployment.",
    "helpLink": "https://konflux-ci.dev/docs/how-tos/creating/",
    "icon": "test-file-stub",
    "title": "Install our GitHub app",
  },
  Object {
    "cta": Object {
      "analytics": Object {
        "app_name": "test-application",
        "link_location": "whats-next",
        "link_name": "view-build-activity",
        "workspace": undefined,
      },
      "href": "/application-pipeline/workspaces/undefined/applications/test-application/activity",
      "label": "View build activity",
    },
    "description": "Make a change to your source code to automatically trigger a new build.",
    "helpLink": "https://konflux-ci.dev/docs/how-tos/creating/",
    "icon": "test-file-stub",
    "title": "Make a code change",
  },
  Object {
    "cta": Object {
      "analytics": Object {
        "app_name": "test-application",
        "link_location": "whats-next",
        "link_name": "manage-build-pipelines",
        "workspace": undefined,
      },
      "label": "Manage build pipelines",
      "onClick": [Function],
    },
    "description": "Add some automation by upgrading your default build pipelines to custom build pipelines.",
    "helpLink": "https://konflux-ci.dev/docs/how-tos/configuring/customizing-the-build/",
    "icon": "test-file-stub",
    "title": "Manage build pipelines",
  },
]
`;
